### ISSUE TITLE
SC '21 | Chimera: Efficiently Training Large-Scale Neural Networks with Bidirectional Pipelines

### ISSUE THREAD

**Issue by vycezhong** [vycezhong/read-papers#222]
https://arxiv.org/pdf/2107.06925.pdf

SC' 21 best paper finalist

**Comment by vycezhong** [comment]
写得非常棒的paper. 可以学习这个写作. 

3.1 key idea就是双向的pipeline叠加. 有两份weights. 这样的memory consumption更加balance. 

这个对比图不错. 
![image](https://user-images.githubusercontent.com/25879526/126972897-e09c449e-b77a-444e-bcc8-5c4ad7b5ee56.png)

3.2 然后分析了通信的策略. 因为有两份weights，要做all-reduce. 下面这种更好. 注意对于中间stage，不应该采用eager gradient synchronization策略. 因为中间的stage空当在后面. 
![image](https://user-images.githubusercontent.com/25879526/126972966-c1e0b7e6-c5a3-4556-a031-51264daa9cfe.png)

3.3讲了如何和data parallelism结合. 很自然. 

3.4讲了如何选择一些参数：B/W/D. 分别是micro-batch size, #replicas, #stages. 对iteration time建模结果如下：
![image](https://user-images.githubusercontent.com/25879526/126973536-21855470-f3a6-4a28-9a80-a51224b3d190.png)

他这里是认为p2p通信不会跟计算overlap. 里面comm_{unoverlapped}指的是all-reduce通信不能被overlapped的比例. 其中，计算时间由microbenchmarks得到，通信时间根据latency-bandwidth cost model建模. 

3.5讲了batch size很大，导致N > D的情况. idea是把一个(N=D)的情况作为一个basic unit，然后concatenate多个basic unit. 如下图. 

![image](https://user-images.githubusercontent.com/25879526/126974730-d3392250-f42d-4698-9440-c819d386ca67.png)

一个降低buble ratio的idea是forward doubling. 就是把两个fw放一起. 还提到一个backward halving, 看上去是减半batch size，没给图. 

3.6 讲了如何拓展到多个pipelines叠加. 看图.
![image](https://user-images.githubusercontent.com/25879526/126975970-c9ebf8d5-5497-40a7-ab21-c673aeefdfd6.png)


实验. workload是BERT-48和GPT-2. 

4.1 mermoy consumption. 主要一个卖点. 因为Chimera的memory consumption更加balance. 
![image](https://user-images.githubusercontent.com/25879526/126976088-30a7e1a4-7c0b-4f35-99c4-25c6038eee09.png)

4.2 parallel scalability. 这个实验先grid search找到baseline的最优配置. 

然后根据3.4的performance model给出预测结果，和实验结果很接近. 但GPT-2的选择有误差. 文中解释是说高估了activation recomputation的时间. 
![image](https://user-images.githubusercontent.com/25879526/126976785-74b2dc0c-07d4-47f9-a253-0da6fa7ee3cb.png)

最后才正式比较各自的最优performance. 
![image](https://user-images.githubusercontent.com/25879526/126976946-044513c5-2112-4eb0-bd3e-0c0dab9d70b1.png)

![image](https://user-images.githubusercontent.com/25879526/126976970-e4b9c316-46c6-43b5-b70e-d643e4f43d56.png)

4.3算是3.5的实验. 同样，每个baseline都做了search parameter space. 看上去的确是同步pipeline里面最好的. 
![image](https://user-images.githubusercontent.com/25879526/126977034-9d565e32-936c-474c-87ab-ec8ce724ae67.png)

4.4是3.6的实验. 多个pipeline叠加. pipeline不是叠加越多就越好，有trade-off的，因为weight数量也会增多，导致显存不够，然后只能降低B，从而降低训练效率. 
![image](https://user-images.githubusercontent.com/25879526/126977606-76bdb545-3030-40f1-a2be-92ae86a7406f.png)
